### ISSUE TITLE
Auth: add OAuth authentication method

### ISSUE THREAD

**Issue by kayrus** [gophercloud/gophercloud#1418]
[API reference](https://developer.openstack.org/api-ref/identity/v3-ext/?expanded=authenticate-with-identity-api-detail#os-oauth1-api)

Keystone code:
- https://github.com/openstack/keystone/blob/master/keystone/auth/plugins/oauth1.py
- https://github.com/openstack/keystone/blob/master/keystone/api/os_oauth1.py

Even using the generated token, there is an issue:

> `*terraform.EvalOpFilter, err: Request forbidden: [POST http://identity:5000/v3/auth/tokens], error message: {"error": {"message": "You are not authorized to perform the requested action: Using OAuth-scoped token to create another token. Create a new OAuth-scoped token instead.", "code": 403, "title": "Forbidden"}}`

```
2019/01/29 12:57:36 [DEBUG] OpenStack Request URL: POST https://identity/v3/auth/tokens
2019/01/29 12:57:36 [DEBUG] OpenStack Request Headers:
Accept: application/json
Content-Type: application/json
User-Agent: Terraform/0.11.7 gophercloud/2.0.0
2019/01/29 12:57:36 [DEBUG] OpenStack Request Body: {
  "auth": {
    "identity": {
      "methods": [
        "token"
      ],
      "token": {
        "id": "***"
      }
    }
  }
}
2019/01/29 12:57:37 [DEBUG] OpenStack Response Code: 403
2019/01/29 12:57:37 [DEBUG] OpenStack Response Headers:
Content-Length: 207
Content-Type: application/json
Date: Tue, 29 Jan 2019 11:57:37 GMT
Server: nginx/1.15.3
Strict-Transport-Security: max-age=15724800; includeSubDomains
Vary: X-Auth-Token
X-Openstack-Request-Id: req-89efad60-a071-4287-83c5-fd3cd8eca1a7

2019/01/29 12:57:37 [DEBUG] OpenStack Response Body: {
  "error": {
    "code": 403,
    "message": "You are not authorized to perform the requested action: Using OAuth-scoped token to create another token. Create a new OAuth-scoped token instead.",
    "title": "Forbidden"
  }
}
```

Ref: #1214

**Comment by kayrus** [comment]
@jtopjian do we really need to create a subtoken for a token, when OS_AUTH_TOKEN is used? Can we just use send it via the header? Is it related to a scope? As far as I understand, when you already have a token, it already has a scope properties assigned, and there is no need to scope it again.

**Comment by jtopjian** [comment]
> As far as I understand, when you already have a token, it already has a scope properties assigned, and there is no need to scope it again.

Unscoped tokens are possible.

I'm not sure what the issue could be here. Token authentication works fine for me when I do:

```shell
$ source openrc
$ openstack token issue
$ export OS_AUTH_TOKEN=...
$ closerc
$ export OS_AUTH_URL="http://keystone:5000/v3/"
$ terraform init
$ terraform apply
```

Perhaps this is something specific to OAuth?

**Comment by kayrus** [comment]
I suppose the current logic is wrong. When we already have a token, provided by user, we should not create yet another token, but use the existing one.

**Comment by jtopjian** [comment]
I think that's a valid and correct process, but the existing implementation must be in place for a reason. It's always possible that it was easier to just create a new token based off of an existing token to handle both cases of an unscoped and scoped token being used, but now a limitation is being discovered when using OAuth.

You're more than welcome to explore changing the implementation.

**Comment by kayrus** [comment]
token issue has been solved in #1752
